Added {get, get_mut, take}_unchecked methods and their benchmarks to Stash

Robbepop opened the pull request:
I have added the methods  `get_unchecked`, `get_unchecked_mut` and `take_unchecked` to `Stash` and also added associated benchmark tests for them. The speedup for `get_unchecked`, `get_unchecked_mut` is quite decent (about 50%) while for `take_unchecked` it isn't so huge (just about 10%). Also a `setup` routine for the `Stash` benchmarks was added.

The use-case for the user knowing when he or she can "safely" use the unchecked versions is similar to `Vec`s `get_unchecked` methods and should be provided by this crate since it is an important API for performance depending projects that could benefit from this data structure such as SAT solvers.

Stebalien commented:
Thanks for the PR.

As-is, these appear to be partially unchecked. They don't check if the index is in bounds but still check if the slot is full. Personally, I would expect them to return the values directly (not wrapped in options) like `slice::get_unchecked` and friends. You can use [unreachable](https://crates.io/crates/unreachable) in the None/Empty cases and should be able to get an even better speedup.

Note: Try benchmarking with both `unreachable()` and `panic!()` to see if there's a difference. Whether or not `unreachable()` helps is very dependent on the LLVM optimizer.

Note 2: `std::unreachable!()` is *NOT* the same thing.

Stebalien commented on benches/stash.rs:
```diff
@@ -27,12 +27,76 @@ fn bench(b: &mut Bencher) {
+    b.iter(|| {
+        let t1 = stash.put("something");
+        let t2 = stash.put("something");
+        let _ = unsafe{ stash.take_unchecked(t1) };
```
You need to use `test::black_box` here to ensure this actually does something. Also, with the *current* API, you should be unwrapping.

Stebalien commented on benches/stash.rs:
```diff
@@ -27,12 +27,76 @@ fn bench(b: &mut Bencher) {
+    let (stash, tickets) = setup();
+    b.iter(|| {
+        for &t in tickets.iter() {
+            test::black_box(stash.get(t));
```
Again, you should probably unwrap to emulate "normal" behavior.

Robbepop replied:
Okay, I will do that. Sorry, didn't see your review earlier.

Stebalien commented on src/stash/mod.rs:
```diff
@@ -364,20 +364,41 @@ impl<V> Stash<V> {
     pub fn take(&mut self, index: usize) -> Option<V> {
-        if let Some(entry) = self.data.get_mut(index) {
-            match mem::replace(entry, Entry::Empty(self.next_free)) {
+        match self.data.get_mut(index) {
```
What was your motivation behind this change? I don't really feel strongly one way or another but it doesn't really fit with the PR (unless I'm missing something).

Robbepop replied:
yeah you are right. It was just a test if Rust compiler was able to go for more optimization if I used a single epression for the entire function body. But there were no differences in performance. Besides that it now looks more uniformely since I implemented the other methods in that style, too. Of course, it is your decision if you want to pull it that way.

Robbepop commented:
I also thought about what API (Option<V> or &V) is more useful and make more sense to users. I think you got a decent point and I think adding another method-cathegory as 'expect' and 'expect_mut' for this use-case isn't a good idea ..

I will implement and benchmark your idea and update this PR as soon as I am done.

Stebalien commented on src/stash/mod.rs:
```diff
@@ -364,20 +364,41 @@ impl<V> Stash<V> {
+    pub unsafe fn take_unchecked(&mut self, index: usize) -> Option<V> {
+        let entry = self.data.get_unchecked_mut(index);
+        match mem::replace(entry, Entry::Empty(self.next_free)) {
+            Entry::Empty(free_slot) => {
```
In the truly unchecked case, you'd mark this case as `unreachable()`. Ditto for the other unchecked functions.

Robbepop commented:
The benchmarks:

With `Option<&V>` as return type:
```
test get               ... bench:         135 ns/iter (+/- 39)
test get_mut           ... bench:         135 ns/iter (+/- 20)
test get_unchecked     ... bench:          86 ns/iter (+/- 24)
test get_unchecked_mut ... bench:          83 ns/iter (+/- 8)
```

With `&V` as return type:
```
test get               ... bench:         136 ns/iter (+/- 10)
test get_mut           ... bench:         137 ns/iter (+/- 23)
test get_unchecked     ... bench:          68 ns/iter (+/- 5)
test get_unchecked_mut ... bench:          71 ns/iter (+/- 22)
```

So using the `reachable` crate we can gain even more speedup as you have expected! :D 
(Thanks for letting me know of this crate btw.!)

Note: std's `unreachable!` macro is useless and even slower than returning Option:
```
test get               ... bench:         138 ns/iter (+/- 40)
test get_mut           ... bench:         138 ns/iter (+/- 39)
test get_unchecked     ... bench:         105 ns/iter (+/- 6)
test get_unchecked_mut ... bench:         125 ns/iter (+/- 30)
```

Stebalien commented:
Nice! Just checking, is that after applying the changes suggested in my comments about unwrapping and `black_box`?

Robbepop commented:
Yes, these tests are before and after your suggested changes for the return type.

I use `blackbox` wherever it was needed, I think.
I haven't yet tested `panic!` against `unreachable()` - will do soon.
The new `take_unchecked` method with added attribute `#[inline]` very much profits from inlining and is much faster now.
(I have also added `#[inline]` to `take` during comparison - I think the optimizer inlines the unchecked version but not the normal one and thus it profits so much from adding `#[inline]`.)

Robbepop commented:
Benchmarks testing `unreachable()` against `panic!()` are indicating that there is an enormous speedup for the `unreachable()` crate function! So we should stick to it.

I have also applied the change from `unreachable!` macro to the `unreachable()` function for the `put` method in `Stash` which leads to minor speed ups as far as I can see.

Robbepop commented:
As far as I can tell your suggested changes have been made:
 - Changed return type of `_unchecked` methods from `Option<V>` to `V`.
 - Used `unreachable()` LLVM intrinsic to further speed up unchecked methods.
 - Additionally:
     * added some new benchmarks
     * improved `Stash::put` performance slightly by making it use the `unreachable` function instead of the macro, too.
     * Benchmarked `panic!()` against `unreachable()`: `unreachable` is faster!

Note: Since the return type of `std::ops::Index` implementation for `Stash` also expects a non-empty bucket, we could apply the same performance improvements to it and use the unchecked versions for it internally.

Robbepop commented:
Current test statistics:

```
test get                          ... bench:         185 ns/iter (+/- 23)
test get_mut                      ... bench:         129 ns/iter (+/- 11)
test get_unchecked                ... bench:          63 ns/iter (+/- 3)
test get_unchecked_mut            ... bench:          63 ns/iter (+/- 31)
test insert_delete                ... bench:           6 ns/iter (+/- 1)
test iter                         ... bench:         102 ns/iter (+/- 42)
test iter_sparse                  ... bench:          86 ns/iter (+/- 35)
test ops_index                    ... bench:         103 ns/iter (+/- 48)
test put_and_take                 ... bench:          23 ns/iter (+/- 3)
test put_and_take_block           ... bench:          84 ns/iter (+/- 14)
test put_and_take_unchecked       ... bench:          21 ns/iter (+/- 5)
test put_and_take_unchecked_block ... bench:          69 ns/iter (+/- 32)
```

Robbepop commented:
Have I gotten it right with `black_box` and `.unwrap()` ?

Robbepop commented:
I think I am done now.  :)

Stebalien commented:
Almost, you need to call `black_box` on the result of *each* call to `take`/`take_unchecked`. Basically, `black_box(value)` forces rust to actually compute `value`. If you just throw the result of `take`/`take_unchecked` away, rustc/LLVM can just optimize them away.

Robbepop commented:
oh, that are many calls to black_box then xD I will do that!

Robbepop commented:
```
test get                          ... bench:         160 ns/iter (+/- 1)
test get_mut                      ... bench:         152 ns/iter (+/- 37)
test get_unchecked                ... bench:          66 ns/iter (+/- 13)
test get_unchecked_mut            ... bench:          68 ns/iter (+/- 18)
test insert_delete                ... bench:           8 ns/iter (+/- 0)
test iter                         ... bench:          97 ns/iter (+/- 4)
test iter_sparse                  ... bench:          88 ns/iter (+/- 20)
test ops_index                    ... bench:         109 ns/iter (+/- 29)
test put_and_take                 ... bench:          33 ns/iter (+/- 0)
test put_and_take_block           ... bench:         122 ns/iter (+/- 1)
test put_and_take_unchecked       ... bench:          24 ns/iter (+/- 1)
test put_and_take_unchecked_block ... bench:          88 ns/iter (+/- 14)
```

Stebalien commented:
Hm. After looking at the diffs, I guess the lack of  `test::black_box`'s is my fault (oops). Sorry about that.

---

A few last things:

* [ ] Does `#[inline]` actually help with `take`? Your comment lead me to believe it doesn't. If it doesn't, it would be best to remove it as over-inlining increases code size and compile times.
* [ ] Unwrap the results in the `get`/`get_mut` benchmarks.

If you feel so inclined, it would be nice to split this into a few logical commits (e.g. refactor `take`,  implement unsafe index functions, replace `unreachable!()` macro in `put` with an unsafe unreachable intrinsic, ...). However, that's not really necessary.

Robbepop commented:
Good take about the `#[inline]` it truely is unnecesary!
Added the `.unwrap`s for `get` and `get_mut` benches, however, no difference in results detected.

Maybe I feel inclined to split the commits into logical commits.
Is it a lot of work? What do I have to do for that? (Sorry, I am kind of new to PRs)

Stebalien commented:
> Good take about the #[inline] it truely is unnecesary!
> Added the .unwraps for get and get_mut benches, however, no difference in results detected.

Thanks!

> Maybe I feel inclined to split the commits into logical commits.
> Is it a lot of work? What do I have to do for that? (Sorry, I am kind of new to PRs)

In this case, I'd recommend using `git reset` to rollback your git log to the commit before your first commit. If you've added my repo as a remote, you can do this by running `git reset my_remote/master`. This *won't* undo your changes, it'll just uncommit them (unless you pass the `--hard` flag; don't do that!).

You can then repeatedly use `git add --patch some_file` to stage the changes you want to put in a commit piece by piece. Once you've staged the set of changes you want to commit, commit them normally and repeat the process for the next set of changes.

Finally, you can force push (`git push -f`) your changes.

Note 1: You can also use `git rebase -i` but, in this case, resetting is probably the best option.
Note 2: If some part (or all) of these instructions are unclear, please ask for clarification. I've left out a lot of explanation of lazyness :sleeping:.

---

Don't worry about accidentally deleting your code; GitHub (and I) have copies.

Robbepop commented:
Okay I will try to do that as soon as I find time for it - maybe on wednesday. Looks like a good learning session for git usage!

Some new benchmarks with a clean system (where no other processes are running):
```
test get                          ... bench:          95 ns/iter (+/- 12)
test get_mut                      ... bench:          93 ns/iter (+/- 8)
test get_unchecked                ... bench:          59 ns/iter (+/- 7)
test get_unchecked_mut            ... bench:          61 ns/iter (+/- 6)
test insert_delete                ... bench:           6 ns/iter (+/- 0)
test iter                         ... bench:          93 ns/iter (+/- 11)
test iter_sparse                  ... bench:          66 ns/iter (+/- 9)
test ops_index                    ... bench:          98 ns/iter (+/- 6)
test put_and_take                 ... bench:          24 ns/iter (+/- 2)
test put_and_take_block           ... bench:          90 ns/iter (+/- 10)
test put_and_take_unchecked       ... bench:          22 ns/iter (+/- 2)
test put_and_take_unchecked_block ... bench:          85 ns/iter (+/- 10)
```

Stebalien commented:
No problem; take your time.

Robbepop committed:
added external dependency unreachable

Robbepop committed:
replaced unreachable! macro invokation with unreachable LLVM intrinsic in put method of Stash

Robbepop committed:
added get_unchecked, get_mut_unchecked and take_unchecked methods to Stash

Robbepop committed:
slightly refactored Stash::take to be single-expression form as the other methods

Robbepop committed:
added a new setup function to Stash benches (not used yet)

Robbepop committed:
made Stash benches use the new setup function

Robbepop committed:
refactored and improved general bench functions for Stash

Robbepop committed:
added bench tests for get, get_mut, get_unchecked and get_mut_unchecked

Robbepop committed:
refactored bench test to benchmark ops::Index impl for Stash

Robbepop committed:
added some put & take combination block benchmarks for Stash

Robbepop commented:
I completely remade the git commits and split them into logical parts.
Hope it is okay now! :)

Stebalien commented:
It's great! Thanks!
